Added warning for not running setup before other functions

ryanxjhan opened the pull request:
## Related Issuse or bug

Info about Issue or bug

Fixes #[2338](https://github.com/pycaret/pycaret/issues/2338)

#### Describe the changes you've made

Added warning for not running setup before other functions. 

## Type of change

Please delete options that are not relevant.
<!--
Example how to mark a checkbox :-
- [x] My code follows the code style of this project.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Pytest

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)

NA

## Checklist:

<!--
Example how to mark a checkbox :-
- [x] My code follows the code style of this project.
-->
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

## Screenshots

 Original           | Updated
 :--------------------: |:--------------------:
 **original screenshot**  | <b>updated screenshot </b> |

ngupta23 commented on pycaret/time_series/forecasting/oop.py:
```diff
@@ -1580,6 +1580,10 @@ def compare_models(
         """
 
+        if not self._setup_ran:
+            raise Exception("setup() must be ran before all other functions. \
```
I would change all `Exceptions` to `RuntimeException`.

ngupta23 replied:
> "setup() must be ran before all other functions"

This is technically not True. You can still run predict_model on a loaded model without running setup. I would reword this as "This function requires the user to run setup() first."

ngupta23 commented on pycaret/time_series/forecasting/oop.py:
```diff
@@ -3972,6 +4011,10 @@ def check_stats(
             Dataframe with the test results
         """
 
+        if not self._setup_ran:
```
Can you remove this for now (just for this method - `check_stats`). `check_stats` is being called from within `setup` before the `_setup_ran` flag is set, hence all unit tests are failing now.

ngupta23 commented:
@Yard1 @ryanxjhan 
I also feel that since the message is the same in all methods, it should be stored as a class attribute. That way, we only type it out once and reuse it in all methods. It will make maintenance easier later if we want to reword this.

ryanxjhan commented:
@ngupta23 Agreed. What about implementing it as a private method?

ngupta23 commented:
> @ngupta23 Agreed. What about implementing it as a private method?

I don't have an issue with that. @Yard1 Do you see any issues?

Yard1 commented on pycaret/time_series/forecasting/oop.py:
```diff
@@ -1700,6 +1704,11 @@ def create_model(
 
         """
+
+        if not self._setup_ran:
```
We should use a decorator first and foremost.

ngupta23 replied:
Seems an overkill and less flexible - don't you think. For example, what if I want to check a few things and only conditionally issue this message (such as in `predict_model`. Having a private method for raising this error would provide more flexibility.

Yard1 replied:
It would be less flexible, but for the PR as it is right now, a decorator would fit the best. But I am fine with a private method also. Whatever we do, however, we should somehow reuse the code.

ngupta23 replied:
O, we went with a private method for now. Currently, it is in the time series module only. Later we may add it to the base class itself.

Yard1 commented:
The best practice here would be to use a decorator instead of editing all the methods

ryanxjhan commented:
Hi @ngupta23, I have implemented the changes. Adding warnings in `predict_model`, `load_model`, and `save_model` does not pass some edge cases in pytest. So I removed the warnings from these methods.

ngupta23 commented:
> Hi @ngupta23, I have implemented the changes. Adding warnings in `predict_model`, `load_model`, and `save_model` does not pass some edge cases in pytest. So I removed the warnings from these methods.

Yes that is correct - these methods do not need the setup to be run.

ngupta23 reviewed (approved):
Changes look good. Falling test is unrelated to time series.
Also movement of this check to base class has been added as a separate issue for now.
